### PR TITLE
Introduce WritePipeline and ReadPipeline objects

### PR DESCRIPTION
**Pull request by joaquincasares**
Hello,

I've added an interface very similar to the `execute_concurrent()` method that is based off the original documentation on how to use callbacks.

The idea is a `Pipeline` object would cache requests and limit results until consumed in an effort to always have a full pipeline of requests. The initial benchmarks seem promising, but perhaps we can do better:

https://github.com/joaquincasares/python-driver/blob/TLP-308/benchmarks/README.md

Example usage of the WritePipeline can be found here: https://github.com/joaquincasares/python-driver/blob/TLP-308/cassandra/concurrent.py#L629-L641

    >>> from cassandra.cluster import Cluster
    >>> from cassandra.concurrent import WritePipeline
    >>> cluster = Cluster(['192.168.1.1', '192.168.1.2'])
    >>> session = cluster.connect()
    >>> write_pipeline = WritePipeline(session)
    >>> prepared_statement = session.prepare("INSERT INTO mykeyspace.users (name, age) VALUES (?, ?)")
    >>> write_pipeline.execute(prepared_statement, ('Jorge', 28))
    >>> write_pipeline.execute(prepared_statement, ('Jose', 28))
    >>> write_pipeline.execute(prepared_statement, ('Sara', 25))
    >>> ...
    >>> write_pipeline.confirm()
    >>> ...
    >>> cluster.shutdown()

Example usage of the ReadPipeline can be found here: https://github.com/joaquincasares/python-driver/blob/TLP-308/cassandra/concurrent.py#L825-L842

    >>> from cassandra.cluster import Cluster
    >>> from cassandra.concurrent import ReadPipeline
    >>> cluster = Cluster(['192.168.1.1', '192.168.1.2'])
    >>> session = cluster.connect()
    >>> read_pipeline = ReadPipeline(session)
    >>> prepared_statement = session.prepare('SELECT * FROM mykeyspace.users WHERE name = ?')
    >>> read_pipeline.execute(prepared_statement, ('Jorge'))
    >>> read_pipeline.execute(prepared_statement, ('Jose'))
    >>> read_pipeline.execute(prepared_statement, ('Sara'))
    >>> prepared_statement = session.prepare('SELECT * FROM old_keyspace.old_users WHERE name = ?')
    >>> read_pipeline.execute(prepared_statement, ('Jorge'))
    >>> read_pipeline.execute(prepared_statement, ('Jose'))
    >>> read_pipeline.execute(prepared_statement, ('Sara'))
    >>> for result in read_pipeline.results():
    ...     for row in result:
    ...         print row.name, row.age
    >>> ...
    >>> cluster.shutdown()

I figured to leave in each individual commit since `git blame` may be able to help if my comments ever do not have context, but don't hesitate to squash the commits if this PR is accepted.

The TL;DR is that by using `pipeline.execute()` we effectively:

* Use `session.execute_async()` without having to manage futures.
* Manage returned future counts to ensure we never run out of memory.
* Handle business logic between sending asynchronous requests and processing them.
* Remove boilerplate `future` code within applications while providing centralized performance enhancements.
* Includes a tuning application within `benchmarks/` to find the number of requests that can saturate without overloading the cluster.

### CODE REVIEW

**Comment by aboudreault** [comment]
Hey @joaquincasares, sorry for the delay. This looks interesting! Could you open a ticket to be sure we schedule and do a follow-up on this. I would also be interested to know more about your use cases for this feature.

**Comment by aboudreault** [comment]
@joaquincasares Hello, is this still something you are working on?

**Comment by joaquincasares** [comment]
Hey @aboudreault , yes, sorry about the delay!

Here's the created ticket: https://datastax-oss.atlassian.net/browse/PYTHON-1046.

Let me know if you have any more questions!

**Comment by absurdfarce** [comment]
Closing due to lack of activity over the past year.  You're free to update the pull request to the current master branch and re-submit if you'd like for us to take another look at this change.
